bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java

```diff
@@ -250,89 +241,6 @@ private void decodeProperties(final BsonReader reader, final DecoderContext deco
         return false;
     }
 
-
-
-    @SuppressWarnings("unchecked")
-    private <S> Codec<S> specializePojoCodec(final PropertyModel<S> propertyModel) {
-        Codec<S> codec = getCodecFromPropertyRegistry(propertyModel);
```

Comment:
This call could loop back to the registry creating new PojoCodec instances. These then could then become specialized instances and weren't cached adequately.

So the specialization and caching were split out to the LazyPropertyModelCodec - which handles the specialization of itself and caches locally on demand.

The PropertyCodecRegistry handles the caching of types per PojoCodec, which can then later be specialized.  This reduces the number of roundtrips to the codecRegistry for types.